mpas_atm refactor to use state structure module

hkershaw-brown opened the pull request:
## Description:
<!--- Describe your changes -->
Refactor mpas_atm model_mod to use state_structure, remove progvar to vector.

### Fixes issue
<!--- link to github issue(s) -->
Various mpas_atm issues, notes in #753
fixes #753 
fixes #523,
fixes mpas part of #389 
fixes #781 
#624, #768

[note](https://docs.google.com/spreadsheets/d/1pNBA0k13k_fbTIGHOF2R1ntk0H1yOq3WDOCTl5theBg/edit?usp=sharing) on subroutines, functions in mpas_atm model_mod refactor vs removed vs unchanged

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Documentation changes needed?
<!-- Put an `x` in all the boxes that apply: -->
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.


### Tests
MPAS model_mod_check and filter

## Checklist for merging

- [ ] Updated changelog entry
- [ ] Documentation updated
- [ ] Update conf.py

## Checklist for release
- [ ] Merge into main
- [ ] Create release from the main branch with appropriate tag
- [ ] Delete feature-branch

## Testing Datasets

- [ ] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [ ] No dataset needed

hkershaw-brown committed:
first pass through removing uneeded code

hkershaw-brown committed:
model_mod_check

hkershaw-brown committed:
get_state_meta_data - bitwise with main

hkershaw-brown committed:
model_interpolate error codes

hkershaw-brown committed:
compute_scalar_with_barycentric with state stucture calls rather than progvar

hkershaw-brown committed:
state variables bounds

Note still have clamp or fail being read but not used #624

hkershaw-brown committed:
model_interpolate

Removed state_vector_to_boundary_file  and update_wind_components so I can compile with debugging flags

hkershaw-brown committed:
get_close and convert_vert for obs and state

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
uncommented find_surrounding_edges

Is the rbf code depricated
 https://github.com/NCAR/DART/issues/753#issuecomment-2442780885

hkershaw-brown committed:
compute_pressure_at_loc

hkershaw-brown committed:
is_edgedata_in_state_vector checks state vector for edge dimensions.

see comment https://github.com/NCAR/DART/issues/753#issuecomment-2444251516

Remove not used variables:
add_static_data_to_diags  - not used
ew_boundary_type, ns_boundary_type  -not used

 Shortened some very verbose comments.

hkershaw-brown committed:
chore: removed unused routine query_vert_localization_coord

https://github.com/NCAR/DART/issues/753#issuecomment-2444418331

removed done todo comment

hkershaw-brown committed:
chore: remove unsused variables from get_state_meta_data

hkershaw-brown committed:
Removed more unused code

statevector_to_boundary_file and statevector_to_analysis file need to be done.

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
note on 2M, 10M qtys

see #768

hkershaw-brown committed:
feat: model_interpolate for qty_vertical_velocity

todo: the pressure on edges

hkershaw-brown committed:
fix: nc can be more than 3 for rbf calculations

set ids array to length nc rather than 3

hkershaw-brown committed:
reject obs to far from surface

Q. Are there any surface obs that are not required for RTTOV?

hkershaw-brown committed:
reject obs above a user specified pressure level

hkershaw-brown committed:
Things that cannot be negative

QTY_CLOUD_FRACTION is also limited to 1

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown commented:
mpas code from MPAS-Model in DART repo is missing the MPAS-Model is missing the [LICENCE](https://github.com/MPAS-Dev/MPAS-Model/blob/master/LICENSE).

hkershaw-brown committed:
chore: add LICENSCE for code from MPAS-Model

moved MPAS-model code to a directory with the LICENCE for the code
from https://github.com/MPAS-Dev/MPAS-Model/blob/master/LICENSE

fixes #781

hkershaw-brown committed:
remove out-of-date revision, revdate

hkershaw-brown committed:
update_mpas_states - copying variables from one file to another

todo: wind, domain_id
see https://github.com/NCAR/DART/issues/753#issuecomment-2445022042
for various notes on update_mpas_states

hkershaw-brown committed:
chore: remove read_2d_from_nc_file

have netcdf utitilies module for this

hkershaw-brown committed:
skipping u for update_mpas_states

update_u_from_reconstruct  passing this namelist option publically.
I am not sure why this is a model_mod option rather than an update_mpas_states
option? To make people set it for filter?
https://github.com/NCAR/DART/issues/753#issuecomment-2562910012

hkershaw-brown committed:
feat: update wind components in update_mpas_states

This was a subroutine update_wind_components in model_mod that was called
by statevector_to_analysis_file

untested

hkershaw-brown committed:
removed used routines and variables, and soon to be unused routines

comment tidy

hkershaw-brown committed:
partial update_bc. Does not compile.

commiting before I rip out the lbc domain

see https://github.com/NCAR/DART/issues/753#issuecomment-2573784421
I think lbc is in the same place as anl_domain

Need to run a regional mpas case to get this info

hkershaw-brown committed:
removed lbc domain

hkershaw-brown committed:
typos and indenting fixes

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
chore: remove svn info

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
use parameter for error code

hkershaw-brown committed:
chore: list the source files that are modified versions of MPAS code

Added this note the the top of the LICENSE included in the
MPAS_model_modules directory

hkershaw-brown commented:
@jlaucar Do you have an estimated timeframe for when you'll be able to review this?

hkershaw-brown committed:
doc: updated model_mod doc namelist info

removed progvar section, removed grid variable list

hkershaw-brown committed:
doc: mpas atm as a search term for the mpas_atm page

hkershaw-brown committed:
remove refernces to clamp from clamp of fail

Fail has been ignored for some time, only clamp is used. #624

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown commented:
@mjs2369 Jeff's not looking at mpas dart code (my bad I though he was working on this), so I'm adding you as a reviewer. 
It is a decent sized pull request, so take a look. Let me know if you want to go over the code and changes, and the original code.

jlaucar commented:
Helen,
I'm going to go ahead and do my best to review this, too. Sorry for
confusion back and forth. I'm very glad that Marlee will be doing a review
which will no doubt be better than mine. Is this something that would be
worth a code review with the three of us sometime? Have a good weekend, Jeff

On Wed, Mar 19, 2025 at 7:45 AM Helen Kershaw ***@***.***>
wrote:

> @mjs2369 <https://github.com/mjs2369> Jeff's not looking at mpas dart
> code (my bad I though he was working on this), so I'm adding you as a
> reviewer.
> It is a decent sized pull request, so take a look. Let me know if you want
> to go over the code and changes, and the original code.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NCAR/DART/pull/780#issuecomment-2736706201>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDHUIUHJMD6ECLLHJYVWQD2VFYIVAVCNFSM6AAAAABTI6UE3OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMZWG4YDMMRQGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: hkershaw-brown]*hkershaw-brown* left a comment (NCAR/DART#780)
> <https://github.com/NCAR/DART/pull/780#issuecomment-2736706201>
>
> @mjs2369 <https://github.com/mjs2369> Jeff's not looking at mpas dart
> code (my bad I though he was working on this), so I'm adding you as a
> reviewer.
> It is a decent sized pull request, so take a look. Let me know if you want
> to go over the code and changes, and the original code.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NCAR/DART/pull/780#issuecomment-2736706201>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDHUIUHJMD6ECLLHJYVWQD2VFYIVAVCNFSM6AAAAABTI6UE3OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMZWG4YDMMRQGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hkershaw-brown commented on models/mpas_atm/model_mod.f90:
```diff
@@ -6129,33 +3263,32 @@ subroutine compute_u_with_rbf(state_handle, ens_size, loc, zonal, uval, ier)
 
-   call find_vert_level(state_handle, ens_size, loc, ncells, celllist, .true., &
-                        lower, upper, fract, ier)
+   call find_vert_level(state_handle, ens_size, loc, ncells, celllist, -1, & !HK @todo this varid needs to signify pressure
```
@hkershaw-brown todo needs addressing with -1

hkershaw-brown replied:
see cec2780ce4ff2baaa63c1cb4388d31864c91e001

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
fix: find_vert_level and on_edge logic for qty & qty not in state

find_vert_level:
- use var_id for vertisheight, maybe on nVertLevelsP1,  nVertLevels
- Use var_id for vertislevel, maybe on nVertLevelsP1,  nVertLevels
- use nVertLevels for vertispressure
- use nVertlevels if obs_qty is not in state

on_edge:
- If qty is not in state on_edge = .false.

Also too_high magic number 81 replaced with too high parameter

mjs2369 commented on models/mpas_atm/MPAS_model_modules/get_coeff_mod.f90:
```diff
@@ -15,10 +10,9 @@ module get_coeff_mod
-character(len=128), parameter :: revdate  = "$Date$"
+character(len=256), parameter :: source   = 'get_coeff_mod'
+character(len=32 ), parameter :: revision = ""
+character(len=128), parameter :: revdate  = ""
```
Here we can remove revision and revdate entirely, like is done in the other files in this PR. Also remove them from the call to the error handler on 262

hkershaw-brown replied:
done in 381b40eeaa98532c2bfb172b88eaa0d02d87cf75

hkershaw-brown committed:
remove svn revision, revdate

Removed save hanging out, module variables are saved.

mjs2369 commented on models/mpas_atm/update_mpas_states.f90:
```diff
@@ -14,83 +14,73 @@ program update_mpas_states
-!         
+!
+!  variables that are not wind, copied from one file to the another
+!  variables that are wind, reconstructed
```
good set of comments but I think the comma is confusing 

"
!  variables that are not wind are copied from one file to the another
!  variables that are wind are reconstructed
"

hkershaw-brown replied:
fixed in 52b87fcae0ef39b320f699c003c66b7f74f6d86

jlaucar commented:
Line 899 in model_mod.f90/qty_ok_to_interpolate
Would it be clearer to roll all the cases that have identical actions into a single case?

Line 167: Why a separate public statement for these three things?

mjs2369 commented:
Few more comment fixes:

[https://github.com/NCAR/DART/blob/381b40eeaa98532c2bfb172b88eaa0d02d87cf75/models/mpas_atm/model_mod.f90#L4330-L4331](https://github.com/NCAR/DART/blob/381b40eeaa98532c2bfb172b88eaa0d02d87cf75/models/mpas_atm/model_mod.f90#L4330-L4331)

R3 is already hardcoded here, think we can remove these lines - maybe just keep the part "comes from the (3d) x/y/z cartesian coordinate" and move it next to where R3 is defined (line 4341 and 4310)

[https://github.com/NCAR/DART/blob/381b40eeaa98532c2bfb172b88eaa0d02d87cf75/models/mpas_atm/model_mod.f90#L1804-L1805](https://github.com/NCAR/DART/blob/381b40eeaa98532c2bfb172b88eaa0d02d87cf75/models/mpas_atm/model_mod.f90#L1804-L1805)

A bit confusing in the wording, is like maybe you should increase MAX_STATE_VARIABLES, but it's not really clear that's what the comment is suggesting. Also, do we want to recommend people change MAX_STATE_VARIABLES? Maybe we just make this a bit bigger now

mjs2369 commented:
Just to clarify:

https://github.com/NCAR/DART/blob/381b40eeaa98532c2bfb172b88eaa0d02d87cf75/models/mpas_atm/model_mod.f90#L902-L917

Is this function (qty_ok_to_interpolate) specifying the QTYs that we can interpolate even if they are not in the state vector?

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
fix: non-zero error for all ensemble members if one fails in find_vert_level

Previously returning early with 'success' for some ensemble members, even
though bailing from the routine.

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
comments: tidy, join case statements that have same result.

hkershaw-brown commented:
> Line 899 in model_mod.f90/qty_ok_to_interpolate Would it be clearer to roll all the cases that have identical actions into a single case?
> 
> Line 167: Why a separate public statement for these three things?

rolled case together & joined public statements in 52b87fcae0ef39b320f699c003c66b7f74f6d864

hkershaw-brown commented:
> Just to clarify:
> 
> https://github.com/NCAR/DART/blob/381b40eeaa98532c2bfb172b88eaa0d02d87cf75/models/mpas_atm/model_mod.f90#L902-L917
> 
> Is this function (qty_ok_to_interpolate) specifying the QTYs that we can interpolate even if they are not in the state vector?

Yup you got it. Some are static, some you need to have other variables.

hkershaw-brown committed:
fix: update_mpas_states to get size of u from the file when not in the state

Note this commit has netcdf fix for #874

hkershaw-brown committed:
remove unused variable- increments

hkershaw-brown committed:
doc: namelist documumentation

match namelist with options in code
removed strange space characters

hkershaw-brown committed:
depricate use_u_for_wind, wind documentation

On advice from jla: error if a user has selected use_u_for_wind
as there is a bug in Manhattan, where all ensemble members are set to
the last ensemble member value in the rbf calculation
https://github.com/NCAR/DART/issues/861
see https://github.com/NCAR/DART/pull/780#issuecomment-2825495746

Documentation for mpas regristry changes needed for DART

Regional wind updates namelist options in update_bc - not sure whey
they do not use the same model_nml options, so not added to docs in
this commit.

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown commented:
still to do regional. /glade/derecho/scratch/hkershaw/DART/MPAS_test_cases_april_2025

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown commented:
NSSL WoF using dart mpas

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
fix: wind variables for update_mpas_states

hkershaw-brown committed:
fix update_bc.f90

skip on lbc_ not u
zonal u, meridonal v
delta_u allocation - only when needed & once

hkershaw-brown committed:
comment update on copying from one file to another

hkershaw-brown committed:
error vs msg if wind options conflict with variables

added todo on hardcoded variables, do all these variables need to be present?

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown committed:
Merge branch 'main' into mpas-refactor

hkershaw-brown commented on models/mpas_atm/model_mod.f90:
```diff
@@ -4732,33 +2170,25 @@ subroutine convert_vert_distrib(state_handle, ens_size, location, obs_kind, ztyp
         zout = 0.0_r8  
         istatus(:) = 0
-        goto 101
+        return
```
needs a set_location before returning

hkershaw-brown commented on models/mpas_atm/model_mod.f90:
```diff
@@ -6576,6 +3716,23 @@ function on_boundary_edge(edgeid)
+!------------------------------------------------------------
+!> Determine if this edge is on the boundary
+
+function cell_next_to_boundary_edge(edgeid)
+
+integer,  intent(in)  :: edgeid
+logical               :: cell_next_to_boundary_edge
+
+if (global_grid .or. .not. allocated(bdyMaskEdge)) return
+
+cell_next_to_boundary_edge = .false.
+cell_next_to_boundary_edge = on_boundary_cell(cellsOnEdge(1,edgeid))
+cell_next_to_boundary_edge = on_boundary_cell(cellsOnEdge(2,edgeid))
+   
+end function cell_next_to_boundary_edge
+
+
```
wonky logic

```suggestion
function cell_next_to_boundary_edge(edgeid)

integer, intent(in) :: edgeid
logical             :: cell_next_to_boundary_edge

cell_next_to_boundary_edge = .false.

if (global_grid .or. .not. allocated(bdyMaskEdge)) return

cell_next_to_boundary_edge = &
   on_boundary_cell(cellsOnEdge(1,edgeid)) .or. &
   on_boundary_cell(cellsOnEdge(2,edgeid))

end function cell_next_to_boundary_edge
```

hkershaw-brown commented on models/mpas_atm/update_bc.f90:
```diff
@@ -27,172 +27,284 @@ program update_bc
+   ! Update variables except 'u' from the analysis
+   allocate(a_var_data(nVertLevels, nCells))
+   allocate(b_var_data(nVertLevels, nCells))
+   allocate(var_data(nVertLevels, nCells))
```
var_data should be initialized to a_var_data?

hkershaw-brown commented on models/mpas_atm/model_mod.f90:
```diff
@@ -6182,30 +3342,28 @@ subroutine compute_u_with_rbf(state_handle, ens_size, loc, zonal, uval, ier)
+  ! Lower/upper could be different levels in pressure
+   do e = 1, ens_size
+      lowindx = get_dart_vector_index(lower(i, e), edgelist(i), dummy,anl_domid, var_id) !HK @todo check x,y,z
+      upindx = get_dart_vector_index(upper(i, e), edgelist(i), dummy,anl_domid, var_id) !HK @todo check x,y,z
```
```suggestion
      lowindx(e) = get_dart_vector_index(lower(i, e), edgelist(i), dummy,anl_domid, var_id) !HK @todo check x,y,z
      upindx(e) = get_dart_vector_index(upper(i, e), edgelist(i), dummy,anl_domid, var_id) !HK @todo check x,y,z
```

hkershaw-brown commented on models/mpas_atm/update_bc.f90:
```diff
@@ -27,172 +27,284 @@ program update_bc
 
 !----------------------------------------------------------------------
 ! Reads lists of input mpas (prior) and filter (analysis) files 
+!HK @todo loop around files, why not run this code in parallel?
```
need to deallocate and reallocate if looping around files.
